File: src/main/java/com/kishlaly/ta/analyze/tasks/blocks/one/Short_ScreenOne_StrictTrendCheck.java
package com.kishlaly.ta.analyze.tasks.blocks.one;

import com.kishlaly.ta.analyze.functions.TrendFunctions;
import com.kishlaly.ta.model.BlockResultJava;
import com.kishlaly.ta.model.SymbolDataJava;
import com.kishlaly.ta.utils.LogJava;

import static com.kishlaly.ta.model.BlockResultCodeJava.NO_DOWNTREND_SCREEN_1;
import static com.kishlaly.ta.model.BlockResultCodeJava.OK;
import static com.kishlaly.ta.analyze.tasks.ThreeDisplays.Config.NUMBER_OF_EMA26_VALUES_TO_CHECK;
import static com.kishlaly.ta.utils.Quotes.resolveMinBarsCount;

/**
 * Strict trend check using quotes, EMA26 and MACD
 */
public class Short_ScreenOne_StrictTrendCheck implements ScreenOneBlock {
    @Override
    public BlockResultJava check(SymbolDataJava screen) {
        boolean downtrendCheckOnMultipleBars = TrendFunctions.downtrendCheckOnMultipleBars(screen, resolveMinBarsCount(screen.timeframe), NUMBER_OF_EMA26_VALUES_TO_CHECK);
        if (!downtrendCheckOnMultipleBars) {
            LogJava.recordCode(NO_DOWNTREND_SCREEN_1, screen);
            LogJava.addDebugLine("No downtrend detected on the long-term screen");
            return new BlockResultJava(screen.getLastQuote(), NO_DOWNTREND_SCREEN_1);
        }
        return new BlockResultJava(screen.getLastQuote(), OK);
    }
}

File: src/main/kotlin/com/kishlaly/ta/analyze/tasks/blocks/one/Short_ScreenOne_StrictTrendCheck.kt
package com.kishlaly.ta.analyze.tasks.blocks.one

import com.kishlaly.ta.analyze.functions.TrendFunctions
import com.kishlaly.ta.analyze.tasks.ThreeDisplays
import com.kishlaly.ta.model.BlockResult
import com.kishlaly.ta.model.BlockResultCode
import com.kishlaly.ta.model.SymbolData
import com.kishlaly.ta.utils.Log
import com.kishlaly.ta.utils.Quotes

class Short_ScreenOne_StrictTrendCheck : ScreenOneBlock {
    override fun check(screen: SymbolData): BlockResult {
        val downtrendCheckOnMultipleBars = TrendFunctions.downtrendCheckOnMultipleBars(
            screen,
            Quotes.resolveMinBarsCount(screen.timeframe),
            ThreeDisplays.Config.NUMBER_OF_EMA26_VALUES_TO_CHECK
        )
        if (!downtrendCheckOnMultipleBars) {
            Log.recordCode(BlockResultCode.NO_DOWNTREND_SCREEN_1, screen)
            Log.addDebugLine("No downtrend detected on the long-term screen")
            return BlockResult(screen.lastQuote, BlockResultCode.NO_DOWNTREND_SCREEN_1)
        }
        return BlockResult(screen.lastQuote, BlockResultCode.OK)
    }
}